File: code/csharp/ch09/luavm/state/LuaValue.cs
using System;
using System.Collections.Generic;
using luavm.api;
using LuaType = System.Int32;

namespace luavm.state
{
    public class LuaValue
    {
//        internal readonly object value;

//        public LuaValue(object value)
//        {
//            this.value = value;
//        }

        public static string toString(object value)
        {
            return Convert.ToString(value);
        }

//        public LuaValue toLuaValue()
//        {
//            return (LuaValue) (value);
//        }

        public static long toInteger(object value)
        {
            return Convert.ToInt64(value);
        }


        public static double toFloat(object value)
        {
            return Convert.ToDouble(value);
        }

        public static bool isString(object value)
        {
            return value.GetType().IsEquivalentTo(typeof(string));
        }

        public static bool isLuaValue(object value)
        {
            return value.GetType().IsEquivalentTo(typeof(LuaValue));
        }

        public static LuaTable toLuaTable(object value)
        {
            return (LuaTable) value;
        }

        public static bool isLuaTable(object value)
        {
            return value.GetType().IsEquivalentTo(typeof(LuaTable));
        }

        public static bool isFloat(object value)
        {
            return value.GetType().IsEquivalentTo(typeof(double));
        }

        public static bool isInteger(object value)
        {
            return value.GetType().IsEquivalentTo(typeof(long));
        }

        internal static LuaType typeOf(object value)
        {
            if (value == null)
            {
                return Consts.LUA_TNIL;
            }

            // Console.WriteLine("\n" + value.GetType().Name);

            switch (value.GetType().Name)
            {
                case "Boolean": return Consts.LUA_TBOOLEAN;
                case "Double": return Consts.LUA_TNUMBER;
                case "Int64": return Consts.LUA_TNUMBER;
                case "String": return Consts.LUA_TSTRING;
                case "LuaTable": return Consts.LUA_TTABLE;
                case "Closure": return Consts.LUA_TFUNCTION;
            }

            throw new Exception("todo!");
        }

        internal static (double, bool) convertToFloat(object value)
        {
            switch (value.GetType().Name)
            {
                case "Double": return (toFloat(value), true);
                case "Int64": return (toFloat(value), true);
                case "String": return number.Parser.ParseFloat(toString(value));
                default: return (0d, false);
            }
        }

        internal static (long, bool) convertToInteger(object val)
        {
            switch (val.GetType().Name)
            {
                case "Int64": return ((long) val, true);
                case "Double": return number.Math.FloatToInteger((double) val);
                case "String": return (Convert.ToInt64(val), true);
                default: return (0L, false);
            }
        }

        private (long, bool) _stringToInteger(string s)
        {
            var v = number.Parser.ParseInteger(s);
            if (v.Item2)
            {
                return (v.Item1, true);
            }

            var v2 = number.Parser.ParseFloat(s);
            if (v2.Item2)
            {
                return number.Math.FloatToInteger(v2.Item1);
            }

            return (0L, false);
        }
    }
}

File: code/python/ch17/src/lua_value.py
from lua_type import LuaType


class LuaValue:
    @staticmethod
    def type_of(val):
        from lua_table import LuaTable
        from closure import Closure
        if val is None:
            return LuaType.NIL
        elif isinstance(val, bool):
            return LuaType.BOOLEAN
        elif isinstance(val, int) or isinstance(val, float):
            return LuaType.NUMBER
        elif isinstance(val, str):
            return LuaType.STRING
        elif isinstance(val, LuaTable):
            return LuaType.TABLE
        elif isinstance(val, Closure):
            return LuaType.FUNCTION
        return LuaType.NONE

    @staticmethod
    def to_boolean(val):
        if val is None:
            return False
        elif isinstance(val, bool):
            return val
        else:
            return True

    @staticmethod
    def is_integer(val):
        return val == int(val)

    @staticmethod
    def parse_integer(s):
        try:
            return int(s, 0)
        except ValueError:
            return None

    @staticmethod
    def parse_float(s):
        try:
            return float(s)
        except ValueError:
            return None

    @staticmethod
    def to_integer(val):
        if isinstance(val, int):
            return val
        elif isinstance(val, float):
            return int(val) if LuaValue.is_integer(val) else None
        elif isinstance(val, str):
            return LuaValue.parse_integer(val)

    @staticmethod
    def to_float(val):
        if isinstance(val, float):
            return val
        elif isinstance(val, int):
            return float(val)
        elif isinstance(val, str):
            return LuaValue.parse_float(val)

    @staticmethod
    def float2integer(val):
        if isinstance(val, float):
            if LuaValue.is_integer(val):
                return int(val)
        return val

    @staticmethod
    def fb2int(val):
        if val < 8:
            return val
        return ((val & 7) + 8) << ((val >> 3) - 1)

    @staticmethod
    def int2fb(val):
        e = 0
        if val < 8:
            return val

        while val >= (8 << 4):
            val = (val + 0xf) >> 4
            e += 4

        while val >= (8 << 1):
            val = (val + 1) >> 1
            e += 1

        return ((e + 1) << 3) | (val - 8)
